include/RAJA/policy/hip/intrinsics.hpp

```diff
@@ -0,0 +1,346 @@
+/*!
+ ******************************************************************************
+ *
+ * \file
+ *
+ * \brief   Header file containing RAJA intrinsics templates for HIP execution.
+ *
+ *          These methods should work on any platform that supports
+ *          HIP devices.
+ *
+ ******************************************************************************
+ */
+
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-24, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#ifndef RAJA_hip_intrinsics_HPP
+#define RAJA_hip_intrinsics_HPP
+
+#include "RAJA/config.hpp"
+
+#if defined(RAJA_ENABLE_HIP)
+
+#include <type_traits>
+
+#include <hip/hip_runtime.h>
+
+#include "RAJA/util/macros.hpp"
+#include "RAJA/util/SoAArray.hpp"
+#include "RAJA/util/types.hpp"
+
+#include "RAJA/policy/hip/policy.hpp"
+
+
+namespace RAJA
+{
+
+namespace hip
+{
+
+namespace impl
+{
+
+/*!
+ * \brief Abstracts access to memory using normal memory accesses.
+ */
+struct AccessorWithFences : RAJA::detail::DefaultAccessor
+{
+  static RAJA_DEVICE RAJA_INLINE void fence_acquire()
+  {
+    __threadfence();
+  }
+
+  static RAJA_DEVICE RAJA_INLINE void fence_release()
+  {
+    __threadfence();
+  }
+};
+
+/*!
+ ******************************************************************************
+ *
+ * \brief Abstracts access to memory using atomic memory accesses.
+ *
+ * \Note Memory access through this class does not guarantee safe access to a
+ *       value that is accessed concurrently by other threads as it may split
+ *       memory operations into multiple atomic instructions.
+ * \Note Fences used through this class only guarantee ordering, they do not
+ *       guarantee visiblity of non-atomic memory operations as it may not
+ *       actually flush the cache.
+ *
+ ******************************************************************************
+ */
+struct AccessorAvoidingFences
```

Comment:
Is this supposed to represent the "opposite" case of the struct above in some sense? Is there a better name for it that would be more intuitive?

Reply:
I agree the name is not great, I based that on how the rocprim internals talk about it. I thought about naming it something to do with atomics, but its a bit confusing there as well as it uses atomics but doesn't actually guarantee atomicity in its memory accesses its more about memory ordering.

Reply:
I changed the name, let me know what you think.